src/r1cs/sparse.rs

```diff
@@ -0,0 +1,259 @@
+//! # Sparse Matrices
+//!
+//! This module defines a custom implementation of CSR/CSC sparse matrices.
+//! Specifically, we implement sparse matrix / dense vector multiplication
+//! to compute the `A z`, `B z`, and `C z` in Nova.
+//!
+//! Notes:
+//!
+//! There are many sparse matrix crates available in rust. However, most of
+//! these crates are do not parallelize, or are not compatiable with field elements.
+//!
+//! For example, [sprs](https://crates.io/crates/sprs/0.2.1) does not parallelize
+//! its sparse matrix / dense vector multiplication. It's traits also require
+//! `num_traits::Zero`, which is not compatiable with `ff::PrimeField`.
+
+use abomonation::Abomonation;
+use abomonation_derive::Abomonation;
+use ff::PrimeField;
+use rayon::prelude::*;
+use serde::{Deserialize, Serialize};
+
+/// CSR format sparse matrix, We follow the names used by scipy.
+/// Detailed explanation here: https://stackoverflow.com/questions/52299420/scipy-csr-matrix-understand-indptr
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, Abomonation)]
+#[abomonation_bounds(where <F as PrimeField>::Repr: Abomonation)]
+pub struct SparseMatrix<F: PrimeField> {
+  #[abomonate_with(Vec<F::Repr>)]
+  /// all non-zero values in the matrix
+  pub data: Vec<F>,
+  /// column indices
+  pub indices: Vec<usize>,
+  /// row information
+  pub indptr: Vec<usize>,
+  /// number of columns
+  pub cols: usize,
+}
+
+impl<F: PrimeField> SparseMatrix<F> {
+  /// construct from Vec<usize(row), usize(col), F>
+  pub fn new(matrix: &[(usize, usize, F)], rows: usize, cols: usize) -> Self {
+    let mut new_matrix = vec![vec![]; rows];
+    for (row, col, val) in matrix {
+      new_matrix[*row].push((*col, *val));
+    }
+
+    for col in new_matrix.iter_mut() {
+      col.sort_by(|(x, _), (y, _)| x.cmp(y));
+    }
+
+    let mut indptr = vec![0; rows + 1];
+    for (i, col) in new_matrix.iter().enumerate() {
+      indptr[i + 1] = indptr[i] + col.len();
+    }
+
+    let mut indices = vec![];
+    let mut data = vec![];
+    for col in new_matrix {
+      let (idx, val): (Vec<_>, Vec<_>) = col.into_iter().unzip();
+      indices.extend(idx);
+      data.extend(val);
+    }
+
+    SparseMatrix {
+      data,
+      indices,
+      indptr,
+      cols,
+    }
+  }
```

Comment:
This structure is effectively implementing a conversion from COO to CSR. The creation of the CSR matrices is done in the method `r1cs_shape`, in `impl NovaShape for MyGroup`, by re-constructing the matrix from constraints. 
Would the benchmark numbers be different if we implemented this in CSR form directly, without the detour through COO?

Reply:
Yes, I'll change this shortly; I was being lazy for the moment.

This conversion isn't in `NIFS::prove` or in any "critical path", so it's a one time cost.